packages/core/src/fhir-deduplication/resources/observation-labs.ts

```diff
@@ -0,0 +1,109 @@
+import { CodeableConcept, Observation } from "@medplum/fhirtypes";
+import { LOINC_CODE, LOINC_OID, SNOMED_CODE, SNOMED_OID } from "../../util/constants";
+import {
+  combineResources,
+  fillMaps,
+  getDateFromResource,
+  pickMostDescriptiveStatus,
+} from "../shared";
+import { extractValueFromObservation, statusRanking } from "./observation-shared";
+
+export function deduplicateObservationsLabs(observations: Observation[]): {
+  combinedObservations: Observation[];
+  refReplacementMap: Map<string, string[]>;
+} {
+  const { observationsMap, refReplacementMap } = groupSameObservationsLabs(observations);
+  return {
+    combinedObservations: combineResources({
+      combinedMaps: [observationsMap],
+    }),
+    refReplacementMap,
+  };
+}
+
+/**
+ * Approach:
+ * 1 map, where the key is made of:
+ * - date
+ * - code
+ * - value
+ */
+export function groupSameObservationsLabs(observations: Observation[]): {
+  observationsMap: Map<string, Observation>;
+  refReplacementMap: Map<string, string[]>;
+} {
+  const observationsMap = new Map<string, Observation>();
+  const refReplacementMap = new Map<string, string[]>();
+
+  function postProcess(
+    master: Observation,
+    existing: Observation,
+    target: Observation
+  ): Observation {
+    const code = master.code;
+    const filtered = code?.coding?.filter(coding => {
+      const system = coding.system?.toLowerCase();
+      return (
+        system?.includes(LOINC_CODE) ||
+        system?.includes(LOINC_OID) ||
+        system?.includes(SNOMED_CODE) ||
+        system?.includes(SNOMED_OID)
+      );
+    });
+    if (filtered) {
+      master.code = {
+        ...code,
+        coding: filtered,
+      };
+    }
+    master.status = pickMostDescriptiveStatus(statusRanking, existing.status, target.status);
+    return master;
+  }
+
+  for (const observation of observations) {
+    const { loincCode, snomedCode } = extractCodes(observation.code);
+    const date = getDateFromResource(observation);
+    const value = extractValueFromObservation(observation);
+
+    if (date && value && (loincCode || snomedCode)) {
```

Comment:
so the approach we said here wasnt to only use loinc code or snomed. It was to check to make sure the code wasnt empty and to otherwise accept whatever code was there. 

https://metriport.slack.com/archives/C06CQC7GQG3/p1724287840702989?thread_ts=1724287692.771049&cid=C06CQC7GQG3